src/main/java/io/micronaut/build/testresources/TestResourcesLifecycleExtension.java

```diff
@@ -0,0 +1,67 @@
+/*
+ * Copyright 2017-2021 original authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micronaut.build.testresources;
+
+import org.apache.maven.AbstractMavenLifecycleParticipant;
+import org.apache.maven.execution.MavenSession;
+import org.apache.maven.model.Build;
+import org.apache.maven.model.Plugin;
+import org.codehaus.plexus.component.annotations.Component;
+import org.codehaus.plexus.util.xml.Xpp3Dom;
+
+import java.util.List;
+import java.util.function.Consumer;
+
+import static io.micronaut.build.testresources.MicronautStopTestResourcesServerMojo.MICRONAUT_TEST_RESOURCES_KEEPALIVE;
+
+/**
+ * A lifecycle extension which determines if the test resources server should
+ * be stopped when the build is complete.
+ */
+@Component(role = AbstractMavenLifecycleParticipant.class, hint = "test-resources")
+public class TestResourcesLifecycleExtension extends AbstractMavenLifecycleParticipant {
+
+    private static final String EXPLICIT_START_SERVICE_GOAL_NAME = "mn:" + MicronautStartTestResourcesServerMojo.NAME;
+
+    @Override
+    public void afterProjectsRead(MavenSession session) {
```

Comment:
I'm wondering if we can add the `<scope>test</scope>` client dependency using a similar mechanism...

Reply:
Sadly nope, no way to dynamically add dependencies via a plugin.

All we could do is to add a new profile in the parent POM that does that (and potentially the enabled property too), and tell the users here to activate the profile instead. 

But that will have to be in every invocation, i.e., `mvn test -Ptest-resources` or `mvn mn:run -Ptest-resources` or `mvn mn:start-testresources-service -Ptest-resources`. I'm uncertain if this is a better user experience.